[Merged by Bors] - refactor(data/finsupp/basic): split `data/finsupp/basic` into three parts

stuart-presnell opened the pull request:
This PR splits the ~2900 lines of `data/finsupp/basic` into more manageable parts:
* the most basic material (~1000 lines) moves to `data/finsupp/defs`
* lemmas about `finsupp.sum` and `finsupp.prod` move to `algebra/big_operators/finsupp`
* the remaining less-used definitions and lemmas remain in `data/finsupp/basic` (~1600 lines)


---
As discussed in [this Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/finsupp.20refactor/near/290992307).

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

stuart-presnell committed:
Add markers

stuart-presnell committed:
Move up some prod/sum lemmas

stuart-presnell committed:
Move marker

stuart-presnell committed:
Add `add_zero_class` markers

stuart-presnell committed:
Move up `add_zero_class` section

stuart-presnell committed:
Move up some instances

stuart-presnell committed:
Move up some more sum/prod lemmas

stuart-presnell committed:
Move up some non-prod lemmas

stuart-presnell committed:
Move down marker past prod lemmas

stuart-presnell committed:
Add `map_range` markers

stuart-presnell committed:
Move up `map_range` section

stuart-presnell committed:
Add `map_domain` and `comap_domain` markers

stuart-presnell committed:
Move up `equiv_congr_left` section (FAILS!)

stuart-presnell committed:
FIX: move `equiv_map_domain_eq_map_domain`

stuart-presnell committed:
Add more markers

stuart-presnell committed:
Move up `cast_finsupp` section; add markers

stuart-presnell committed:
Edit markers

stuart-presnell committed:
Move down §§ `map_range` & `equiv_congr_left`

stuart-presnell committed:
Move down § `graph`

stuart-presnell committed:
Move 2nd half of file to new `finsupp/misc.lean`

stuart-presnell committed:
Move sum and prod lemmas to `big_operators/...`

stuart-presnell committed:
Fix imports

stuart-presnell committed:
Fix import

stuart-presnell committed:
Fix import

stuart-presnell committed:
Fix import

stuart-presnell committed:
Rename `.../basic` to `.../defs`

stuart-presnell committed:
Rename `.../misc` to `.../basic`

stuart-presnell committed:
Update imports to new filenames

Vierkantor commented on src/data/finset/finsupp.lean:
```diff
@@ -6,6 +6,7 @@ Authors: Yaël Dillies
 import data.finset.pointwise
 import data.finsupp.indicator
 import data.fintype.card
+import algebra.big_operators.finsupp
```
```suggestion
import algebra.big_operators.finsupp
import data.finset.pointwise
import data.finsupp.indicator
import data.fintype.card
```

Vierkantor commented on src/data/finsupp/multiset.lean:
```diff
@@ -4,6 +4,7 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Johannes Hölzl
 -/
 import data.finsupp.order
+import data.finsupp.basic
```
```suggestion
import data.finsupp.basic
import data.finsupp.order
```

Vierkantor commented on src/set_theory/cardinal/ordinal.lean:
```diff
@@ -7,6 +7,7 @@ Authors: Johannes Hölzl, Mario Carneiro, Floris van Doorn
 import order.bounded
 import set_theory.ordinal.principal
 import tactic.linarith
+import data.finsupp.basic
```
```suggestion
import data.finsupp.basic
import order.bounded
import set_theory.ordinal.principal
import tactic.linarith
```

Vierkantor commented on src/algebra/big_operators/finsupp.lean:
```diff
@@ -15,13 +15,479 @@ import algebra.big_operators.order
+
+namespace finsupp
+
+
```
```suggestion
```

Vierkantor reviewed (approved):
Since there doesn't seem to be any objection and the split looks sensible, I'd propose going forward with the current distribution of lemmas, and later we can always rearrange further.

stuart-presnell committed:
Edit imports

stuart-presnell committed:
Merge branch 'master' into SP_finsupp_refactor

stuart-presnell committed:
Delete moved material; fix namespaces

Some material was restored in resolving the merge conflict, which has now been deleted from this file.

stuart-presnell committed:
Delete unneeded imports

stuart-presnell committed:
Restore import of `data.finset.preimage`

stuart-presnell committed:
Delete `finsupp.` prefix

stuart-presnell committed:
Edit header docstring for `finsupp/defs`

stuart-presnell committed:
Move up section `erase`

stuart-presnell committed:
Delete dividers

stuart-presnell committed:
Edit docstrings

stuart-presnell committed:
Remove dividers

stuart-presnell committed:
Remove dividers; edit docstrings

stuart-presnell committed:
Move `prod_dvd_prod_of_subset_of_dvd`

stuart-presnell committed:
Fix error in `totient`

stuart-presnell committed:
Merge branch 'master' into SP_finsupp_refactor

stuart-presnell committed:
Move `zero_update` to defs

stuart-presnell committed:
Delete moved material
